### ISSUE TITLE
YouTube search result list remains empty

### ISSUE THREAD

**Issue by RJVB** [zaps166/QMPlay2#358]
I'm no longer seeing any results of YouTube searches. It does look like a search is being carried out, and I do see search completion suggestions (which might just be cached, of course), but the result list remains empty.

Noticed this today in a build that used to work in this respect but can't say when it might have started. Issue remains with the current HEAD and after updating youtube-dl.

**Comment by zaps166** [comment]
![qmplay2](https://user-images.githubusercontent.com/6239090/94806099-b690af80-03ed-11eb-9c86-eb4d598b8972.gif)

**Comment by taktiker12** [comment]
> I'm no longer seeing any results of YouTube searches. It does look like a search is being carried out, and I do see search completion suggestions (which might just be cached, of course), but the result list remains empty.
> 
> Noticed this today in a build that used to work in this respect but can't say when it might have started. Issue remains with the current HEAD and after updating youtube-dl.

just update qmplay2 .it happened to me too

**Comment by RJVB** [comment]
>just update qmplay2 .it happened to me too

```
> QMPlay2 --help
Usage: /opt/local/bin/QMPlay2 [options] [url]
QMPlay2 - Qt Media Player 2 (20.07.04-git-98db50ec)
```

Already updated, without success...

Wait, I don't get results for, for instance, "Doug Macleod", but I do get results for "doug macleod come to find". "Lag TN270" also returns 0 results.

Typing "test" as in zaps166's reply also gives me results, but not the same as his. That's probably youtube/google being clever - I guess it would be nice if we could specify a preferred language (or indeed, specify explicitly that you want any language).

I've got my preferred resolution set to 720p btw.

At least searching for a video UUID (e.g. NwUEHBmrWkc) works again.

**Comment by zaps166** [comment]
Maybe it depends on localization (IP address), hmm...

![Screenshot_20201001_161948](https://user-images.githubusercontent.com/6239090/94821517-f19cde00-0401-11eb-8abf-35c32bb44506.png)

**Comment by RJVB** [comment]
>Maybe it depends on localization (IP address), hmm...
>
>![Screenshot_20201001_161948](https://user-images.githubusercontent.com/6239090/94821517-f19cde00-0401-11eb-8abf-35c32bb44506.png)

THose sure look like the hits I used to get before (and still get in a browser - at least when logged in).

Speaking of which: YouTube has taken to popping up an in-browser "invitation" to log in which is annoying me in no small way because it appears after the video auto-started playing (and stops+resets playback). 

BTW:

```
> ~/.qmplay2/youtube-dl "ytsearch:doug macleod"
[youtube:search] query "doug macleod": Downloading page 1
[download] Downloading playlist: doug macleod
[youtube:search] playlist doug macleod: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[youtube] Th25d_PCfKw: Downloading webpage
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[download] Destination: Doug MacLeod - The Night of the Devil's Road (Bing Lounge)-Th25d_PCfKw.f137.mp4
[download] 100% of 134.65MiB in 00:26
[download] Destination: Doug MacLeod - The Night of the Devil's Road (Bing Lounge)-Th25d_PCfKw.f251.webm
[download] 100% of 5.63MiB in 00:00
[ffmpeg] Merging formats into "Doug MacLeod - The Night of the Devil's Road (Bing Lounge)-Th25d_PCfKw.mkv"
Deleting original file Doug MacLeod - The Night of the Devil's Road (Bing Lounge)-Th25d_PCfKw.f137.mp4 (pass -k to keep)
Deleting original file Doug MacLeod - The Night of the Devil's Road (Bing Lounge)-Th25d_PCfKw.f251.webm (pass -k to keep)
[download] Finished downloading playlist: doug macleod
```

**Comment by RJVB** [comment]
Any progress on this? Here it seems the issue has become worse; I'm rarely getting any hits at all anymore.

Is there a way to get a terminal dump of youtube's reply, or possibly even of the entire dialogue to see if something glitches there?

**Comment by zaps166** [comment]
Probably youtube has changed once again :/

**Comment by thubble** [comment]
It looks like Google added a single space at the end of the search results JSON definition, which is confusing the parser. This simple patch fixes it, at least for me:

```diff
diff --git a/src/modules/Extensions/YouTube.cpp b/src/modules/Extensions/YouTube.cpp
index 0e859608..31f5fbe5 100644
--- a/src/modules/Extensions/YouTube.cpp
+++ b/src/modules/Extensions/YouTube.cpp
@@ -1080,7 +1080,7 @@ QJsonDocument YouTube::getYtInitialData(const QByteArray &data)
     if (idx2 < 0)
         return QJsonDocument();
 
-    auto jsonData = data.mid(idx, idx2 - idx);
+    auto jsonData = data.mid(idx, idx2 - idx).trimmed();
     if (jsonData.endsWith(';'))
         jsonData.chop(1);
 
```

**Comment by RJVB** [comment]
Not 100% for me, sadly. Some results is better than none, but maybe we can improve the fix even more ... or else detect an error condition and inform the user.

I've modified the end of the same function to

    auto ret = QJsonDocument::fromJson(jsonData);
    qCWarning(youtube) << data << "->" << jsonData << "->" << ret;
    return ret;

and will attach the output of a failed search and of a succeeded search.

**Comment by RJVB** [comment]
output of 

```
    auto ret = QJsonDocument::fromJson(jsonData);
    qCWarning(youtube) << data << "->" << jsonData << "->" << ret;
    return ret;
```

for a failed search and for a succeeding search:
[ytsearch.txt](https://github.com/zaps166/QMPlay2/files/5471522/ytsearch.txt)

**Comment by RJVB** [comment]
Is parsing of the 1st reply failing because of ads?!

**Comment by zaps166** [comment]
@thubble Thanks, you can do PR next time!

**Comment by zaps166** [comment]
Patch: 77cf66c0dd0b032db49ca391ef15842c2b4151b1

**Comment by zaps166** [comment]
> Any progress on this? Here it seems the issue has become worse; I'm rarely getting any hits at all anymore.

Please check again with a fix: 193530399a8b5167180ec5d86f8d3dc2e398729f

**Comment by DarkWav** [comment]
It works great for me, everything returned to normal:
![Screenshot_QMPlay2](https://user-images.githubusercontent.com/16162641/98422060-0a488580-208b-11eb-9759-f48e0532df44.png)

**Comment by zaps166** [comment]
@DarkWav 

>  The only thing that's awkward is that search result pages no longer work, it always displays the first page for some reason.

Because YouTube no longer has pages, it loads new results when scrolling down - it is not implemented yet here. Now wait when Google browser removes pages and we'll be unable to find anything :-1:

**Comment by RJVB** [comment]
On Saturday November 07 2020 04:13:44 Błażej Szczygieł wrote:
>@DarkWav 
>
>>  The only thing that's awkward is that search result pages no longer work, it always displays the first page for some reason.
>
>Because YouTube no longer has pages, it loads new results when scrolling down - it is not implemented yet here. Now wait when Google browser removes pages and we'll be unable to find anything :-1: 

and all that probably because of all those people browsing from a phone... (but even there I hate it because try to jump back to the top without scrolling (or back down if ever the page reloads for some reason).

Or maybe it's to force more ads onto us?

**Comment by DarkWav** [comment]
@zaps166 thanks for the explanation. I guess I'll just have to enter more exact search terms now.

**Comment by zaps166** [comment]
> and all that probably because of all those people browsing from a phone... 

The did the same for desktop website which is inconvenient. Also inconvenient to me on mobile phone :laughing: I have to check how it works and add to QMPlay2 later.

---

I rarely use YouTube, because this website is not good for music (terrible quality is some cases, medium quality in most cases), also not good for movies. It is good only for home made videos, so I can use web browser in this case (especially when I have text message with an URL)... Moreover, ads everywhere, Chrome + uBlock has adds there, only Firefox + uBlock blocks this properly. And now this annoying login to YouTube popup and even more things to accept about privacy...

Also this: https://github.com/github/dmca/blob/master/2020/10/2020-10-23-RIAA.md

So I stopped checking if YouTube still works in QMPlay2...

---

For music there are many other better services (torrent or not) like Deezer, Tidal (both can stream FLAC). Even Spotify (desktop app only, not website) is much better than YouTube, or MyFreeMP3.

For movies - also there are many better websites (legal or not) + Netflix.

**Comment by RJVB** [comment]
On Saturday November 07 2020 12:27:36 Błażej Szczygieł wrote:

>The did the same for desktop website which is inconvenient.

What I meant is that all those people who think their phone is the ultimate computing device want to feel at home too if they are forced to use a desktop computer...

>I rarely use YouTube, because this website is not good for XXXX

There's still a lot of content there from content makers that is much more convenient to watch in a proper playback app that in the webbrower, for just about all reasons one can think of (including not having a current, high-end PC). Tutorials, cooking videos (yeah, I watch those), music lessons and these days lots live streamed concerts. There is enough music content that has a decent enough quality that's more or less on par with what you can get over FM radio.
>
>Also this: https://github.com/github/dmca/blob/master/2020/10/2020-10-23-RIAA.md

I guess we'll see how that plays out. Microsoft is certainly big enough to shrug this off with the message that they're just hosting the software, not using it. There's probably a lot of source code being hosted on github that can be used to do much more illegal things than thwarting copyright...

>So I stopped checking if YouTube still works in QMPlay2...

Don't worry, we can take care of that ;)

**Comment by zaps166** [comment]
> The only thing that's awkward is that search result pages no longer work, it always displays the first page for some reason.

Fixed: 75369135422b060174ed8b87abbb54e026d4771e

**Comment by DarkWav** [comment]
Thanks! It works great!
